File: graph_algorithms/bfs.cpp
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

vector <vector<int>> graph;
vector <bool> flag;

void connect_edge(int from, int to){
    graph[from].push_back(to);
}

void bfs(int start_node){
    queue <int> q;

    q.push(start_node); // add the start node to start visit
    flag[start_node] = true; // mark the fast node as visited

    while(!q.empty()){
        int front_node = q.front();
        q.pop(); // remove the top node

        cout<< front_node << endl;

        // add all adjacent nodes to the queue
        for(vector<int> :: iterator i = graph[front_node].begin(); i != graph[front_node].end(); i++){
            if(!flag[*i]){
                q.push(*i);
                flag[*i] = true; // mark it as visited
            }
        }
    }
}

int main(){
    int nodes = 4;

    flag.assign(nodes, false);
    graph.assign(nodes, vector<int>());

    connect_edge(0,1);
    connect_edge(0,2);
    connect_edge(1,2);
    connect_edge(2,1);
    connect_edge(3,2);
    connect_edge(2,3);

    bfs(0);

    return 0;
}

File: graph_algorithms/bfs.py
graph = {}
visited = []  # keep track of visited nodes

def add_edge(a, b):
    if a in graph.keys():
        graph[a].append(b)
    else:
        graph[a] = [b]
    
    # if bidirectional add other also



def bfs(start_node):
    queue = []

    queue.append(start_node)
    visited.append(start_node)

    while queue:
        front_node = queue.pop(0)

        print(front_node)

        for n in graph[front_node]:
            if n not in visited:
                visited.append(n)
                queue.append(n)


add_edge(0,1)
add_edge(0,2)
add_edge(1,2)
add_edge(2,1)
add_edge(3,2)
add_edge(2,3)

print(graph)
bfs(0)